posix: standardize kconfig options

cfriedt opened the pull request:
This change (outlined as part of #51211) deprecates several POSIX Kconfig variables (linked below) in favour of their more normative replacements defined by IEEE 1003.1-2017.

[Release Notes](https://builds.zephyrproject.io/zephyr/pr/73047/docs/releases/release-notes-3.7.html)

[Doc Preview](https://builds.zephyrproject.io/zephyr/pr/73047/docs/services/portability/posix/index.html)

Out-of-tree users of the POSIX API should use the provided script to automatically update to the new Kconfig options.
```shell
$ python $ZEPHYR_BASE/scripts/utils/migrate_posix_kconfigs.py -r root_path
```

See also
[Migration Guide](https://builds.zephyrproject.io/zephyr/pr/73047/docs/releases/migration-guide-3.7.html)

- [x] `DEPRECATED` added to older symbols following [policy](https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#deprecated)
- [x] add section / script in migration guide
- [x] email sent to zephyr-devel

Fixes #66132

Commit:
posix: unistd: move posix features to separate header file

Move posix features to a separate header file and include that
header file from unistd.h.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: add Kconfig.deprecated

There will be some POSIX Kconfig symbols that will be deprecated
in favour of Kconfig symbols that are exact matches for those in
the POSIX specification (1003.1-2017).

Create the empty Kconfig file now, and then incrementally add to
it as non-standard POSIX Kconfig options are replaced with
standard-like POSIX Kconfig options.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: timers: deprecate CONFIG_POSIX_CLOCK and TIMER

The POSIX_CLOCK option does not correspond to any standard
option. It was used to active features of several distinct
POSIX Options and Option Groups, which complicated API and
application configuration as a result.

POSIX_CLOCK is being deprecated in order to ensure that Zephyr's
POSIX Kconfig variables correspond to those defined in the
specification, as of IEEE 1003.1-2017.

Additionally, CONFIG_TIMER is being deprecated because it does
not match the corresponding POSIX Option (_POSIX_TIMERS).

With this deprecation, we introduce the following Kconfig
options that map directly to standard POSIX Option Groups by
simply removing "CONFIG_":

* CONFIG_POSIX_TIMERS

Similarly, we introduce the following Kconfig options that
map directly to standard POSIX Options by simply removing
"CONFIG":

* CONFIG_POSIX_CLOCK_SELECTION
* CONFIG_POSIX_CPUTIME
* CONFIG_POSIX_DELAYTIMER_MAX
* CONFIG_POSIX_MONOTONIC_CLOCK
* CONFIG_POSIX_TIMEOUTS
* CONFIG_POSIX_TIMER_MAX

In order to maintain parity with the current feature set, we
introduce the following Kconfig options that map directly to
standard POSIX Option Groups by simply removing "CONFIG_":

* CONFIG_POSIX_MULTI_PROCESS - sleep()

Similarly, in order to maintain parity with the current feature
set, we introduce the following additional Kconfig options that
map directly to standard POSIX Options by simply removing
"CONFIG":

* CONFIG_XSI_SINGLE_PROCESS - gettimeofday()

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: features: leave feature test macros undefined if unimpl

Rather than forcing the macro to -1, simply leave it undefined
if the particular feature in question is unimplemented. This is
the convention used by external libcs.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: sysconf: correct typo on _SC_MAPPED_FILE(S)

The standard macro is plural - so _SC_MAPPED_FILES - which
corresponds to _POSIX_MAPPED_FILES.

In Zephyr, we previously had _SC_MAPPED_FILE and
_POSIX_MAPPED_FILE, so it's a simple correction.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: sysconf: define z sysconf macros with COND_CODE_1

It's sort of more conventient to do the opposite of what was
done with the __z_sysconf_ macros, and use them for both the
full implementation and the macro implementation.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: signals: deprecate CONFIG_POSIX_LIMITS_RTSIG_MAX

There were two Kconfig constants defined for (effectively) the
same thing in Zephyr. Namely, RTSIG_MAX - the highest valued
real-time signal.

Deprecate CONFIG_POSIX_LIMITS_RTSIG_MAX in favour of
CONFIG_POSIX_RTSIG_MAX since the latter more closely matches
convention.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
tests: posix: headers: remove tests for varying features

The header tests for the existence of constants like
_POSIX_SPIN_LOCKS are bound to fail in one circumstance or
another because each of those options are (theoretically)
features which can be enabled or disabled.

Moreover, some implementations (e.g. external libc) might
not implement one particular subset or option group.

These tests are poorly conditioned, so remove them.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
tests: posix: headers: check some structs only when posix_api=y

Some POSIX header tests assumed that various structures would be
declared even if POSIX_API=n. Actually most meaningful POSIX
structures are guarded by feature test macros defined by
the implementation (i.e. implementation conformance macros).

So only run tests on those structures when CONFIG_POSIX_API=y.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
tests: posix: headers: enable checks for sysconf() constants

For some reason, even though sysconf() has multiple
implementations in Zephyr, the header tests were not checking
for the existence of _SC_ constants.

Uncomment the checks - they should be run on a regular basis.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
tests: posix: headers: enable checks for confstr() constants

For some reason, even though confstr() is implemented in Zephyr,
the header tests were not checking for the existence of _CS_
constants.

Uncomment the tests so that they can be run regularly.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

ycsin commented on lib/posix/options/Kconfig.signal:
```diff
@@ -28,13 +36,6 @@ config POSIX_SIGNAL_STRING_DESC
 	  Use full description for the strsignal API.
 	  Will use 256 bytes of ROM.
 
-config POSIX_LIMITS_RTSIG_MAX
-	int "_POSIX_RTSIG_MAX value in limits.h"
-	default 8
-	help
-	  Define the _POSIX_RTSIG_MAX value in limits.h.
-	  IEEE 1003.1 defines this to be 8.
-
```
I think this was implementing `_POSIX_RTSIG_MAX`:

https://pubs.opengroup.org/onlinepubs/7908799/xsh/limits.h.html
```
RTSIG_MAX
Maximum number of realtime signals reserved for application use in this implementation. Minimum Acceptable Value: _POSIX_RTSIG_MAX
```
```
_POSIX_RTSIG_MAX
The number of realtime signal numbers reserved for application use. Value: 8
```

cfriedt replied:
This way uses one fewer Kconfig variable and puts the default minimum in Kconfig.

ycsin commented on lib/posix/options/signal.c:
```diff
@@ -13,8 +13,7 @@
 
-BUILD_ASSERT(CONFIG_POSIX_LIMITS_RTSIG_MAX >= 0);
-BUILD_ASSERT(CONFIG_POSIX_RTSIG_MAX >= CONFIG_POSIX_LIMITS_RTSIG_MAX);
+BUILD_ASSERT(CONFIG_POSIX_RTSIG_MAX >= CONFIG_POSIX_RTSIG_MAX);
```
Should this be `>= 8`?

cfriedt replied:
It is 8 by default.

ycsin replied:
Ah.. it doesn't always have to be 8, probably this should be changed to
```suggestion
BUILD_ASSERT(CONFIG_POSIX_RTSIG_MAX >= 0);
```

otherwise it would be comparing against itself and is always true

ycsin commented on include/zephyr/posix/posix_features.h:
```diff
@@ -0,0 +1,217 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
```
Maybe can guard this file with something like:
```c
#if !defined(ZEPHYR_INCLUDE_POSIX_UNISTD_H_) && !defined(ZEPHYR_INCLUDE_POSIX_SIGNAL_H_)
#warning This is a private header file, do not include it directly.
#endif
```

cfriedt replied:
It's not intended to be private.

This will probably be renamed to `sys/features.h` later, which should match newlib / picolibc.

ycsin commented on lib/posix/options/Kconfig.timer:
```diff
@@ -1,13 +1,76 @@
+	bool "POSIX per-process CPU-time clocks"
+	default y if POSIX_API
+	help
+	  This enables CLOCK_MONOTONIC.
```
looks like copy paste oops

cfriedt commented on tests/posix/headers/src/dirent_h.c:
```diff
@@ -19,22 +19,22 @@
+	/* zassert_not_null(scandir); */   /* not implemented */
+	/* zassert_not_null(seekdir); */   /* not implemented */
+	/* zassert_not_null(telldir); */   /* not implemented */
+#endif
```
Oddly, even just including the header without `CONFIG_POSIX_API` is a successful test, in this case.

cfriedt commented:
Just kicking CI again. It looks fine locally for me.

Commit:
posix: deprecate POSIX_MAX_FDS and add POSIX_DEVICE_IO

The POSIX_MAX_FDS option does not correspond to any standard
POSIX option. It was used to define the size of the file
descriptor table, which is by no means exclusively used by
POSIX (also net, fs, ...).

POSIX_MAX_FDS is being deprecated in order to ensure that
Zephyr's POSIX Kconfig variables correspond to those defined in
the specification, as of IEEE 1003.1-2017. Namely,
POSIX_OPEN_MAX. CONFIG_POSIX_MAX_OPEN_FILES is being deprecated
for the same reason.

To mitigate any possible layering violations, that option is
not user selectable. It tracks the newly added
CONFIG_ZVFS_OPEN_MAX option, which is native to Zephyr.

With this deprecation, we introduce the following Kconfig
options that map directly to standard POSIX Option Groups by
simply removing "CONFIG_":

* CONFIG_POSIX_DEVICE_IO

Similarly, with this deprecation, we introduce the following
Kconfig options that map directly to standard POSIX Options by
simply removing "CONFIG":

* CONFIG_POSIX_OPEN_MAX

In order to maintain parity with the current feature set, we
introduce the following Kconfig options.

* CONFIG_POSIX_DEVICE_IO_ALIAS_CLOSE
* CONFIG_POSIX_DEVICE_IO_ALIAS_OPEN
* CONFIG_POSIX_DEVICE_IO_ALIAS_READ
* CONFIG_POSIX_DEVICE_IO_ALIAS_WRITE

Gate open(), close(), read(), and write() via the
CONFIG_POSIX_DEVICE_IO Kconfig option and move
implementations into device_io.c, to be conformant with the
spec.

Lastly, stage function names for upcoming ZVFS work, to be
completed as part of the LTSv3 Roadmap (e.g. zvfs_open(), ..).

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: fd_mgmt: add fd_mgmt option group

With this change, we collect implementations of functions that are
part of the POSIX_FD_MGMT Option Group into one compilation unit
(fd_mgmt.c) and provide a matching Kconfig symbol that is
reflective of the standard.

Additionally, provide Kconfig options for aliases of those functions
since Newlib-based toolchains internally use an underscore-prefixed
alias for many POSIX functions (in this case _lseek).

Prefix Zephyr-native implementations with zvfs_ to prepare for the
up-coming changes in that area.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: deprecate PTHREAD_BARRIER for POSIX_BARRIERS

The standard specifies _POSIX_BARRIERS as an Option so let's
use CONFIG_POSIX_BARRIERS for consistence.

This change deprecates CONFIG_PTHREAD_BARRIER. Users should
instead choose CONFIG_POSIX_BARRIERS.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: deprecate POSIX_ENV _CONFSTR _SYSCONF and _UNAME

This change deprecates the following Kconfig options
* CONFIG_POSIX_ENV
* CONFIG_POSIX_CONFSTR
* CONFIG_POSIX_SYSCONF
* CONFIG_POSIX_UNAME

and instead groups them into a single Kconfig option that maps
directly to the standard. Namely, CONFIG_POSIX_SINGLE_PROCESS.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: deprecate PTHREAD_SPINLOCK for POSIX_SPIN_LOCKS

This change deprecates CONFIG_PTHREAD_SPINLOCK in favour of
CONFIG_POSIX_SPIN_LOCKS, which maps directly to the name of the
standard POSIX Option Group.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: deprecate POSIX_MQUEUE in favour of POSIX_MESSAGE_PASSING

This change deprecates CONFIG_POSIX_MQUEUE in favour of
CONFIG_POSIX_MESSAGE_PASSING, which maps directly to the name of the
standard POSIX Option.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: deprecate PTHREAD_RWLOCK for POSIX_READER_WRITER_LOCKS

This change deprecates CONFIG_PTHREAD_RWLOCK in favour of
CONFIG_POSIX_READER_WRITER_LOCKS, which maps directly to the
name of the standard POSIX Option.

Annoyingly, the POSIX_RW_LOCKS Option Group is inconsistently
named. However, it is more convenient for us to use the Option
name since it is also the format used by the sysconf() variable
(_SC_READER_WRITER_LOCKS).

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: deprecate POSIX_FNMATCH GETOPT GETENTROPY

The functions fnmatch(), getopt(), getentropy()
and others are grouped into the standard Option Group
POSIX_C_LIB_EXT.

The getentropy() function is currently in-draft for
Issue 8 as of 2021.

https://www.opengroup.org/austin/docs/austin_1110.pdf

Not surprisingly, the POSIX_C_LIB_EXT Option Group
also includes the highly debated strnlen() function.

Moving that function will be deferred until later.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: deprecate POSIX_FS for POSIX_FILE_SYSTEM

This change deprecates CONFIG_POSIX_FS in favour of
CONFIG_POSIX_FILE_SYSTEM, which maps directly to the name of the
standard POSIX Option Group.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: deprecate SEM_VALUE_MAX and SEM_NAMELEN_MAX

This change deprecates CONFIG_SEM_VALUE_MAX in favour of
CONFIG_POSIX_SEM_VALUE_MAX which maps directly to
_POSIX_SEM_VALUE_MAX.

Additionally, we add the Kconfig option
CONFIG_POSIX_SEMAPHORES which maps directly to the
POSIX Option _POSIX_SEMAPHORES.

For consistence, deprecate CONFIG_SEM_NAMELEN_MAX
in favour of CONFIG_POSIX_SEM_NAMELEN_MAX.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
doc: posix: add kconfig link for async io

Update documentation so that there is a link to
CONFIG_POSIX_ASYNCHRONOUS_IO as it was previously missing.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: deprecate POSIX_SYSLOG for XSI_SYSTEM_LOGGING

This change deprecates the CONFIG_POSIX_SYSLOG Kconfig
variable for CONFIG_XSI_SYSTEM_LOGGING, which has a
direct mapping to the POSIX Option Group XSI_SYSTEM_LOGGING.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: deprecate POSIX_PUTMSG for XOPEN_STREAMS

This change deprecates CONFIG_POSIX_PUTMSG and introduces a new
Kconfig option CONFIG_XOPEN_STREAMS which maps directly to the
standard POSIX Option, _XOPEN_STREAMS.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: deprecate PTHREAD_IPC _MUTEX _COND and _KEY

Zephyr's POSIX API is moving toward using the standard
nomenclature from IEEE 1003.1-2017 for as much as possible. In
particular, we want to have consistent naming between Zephyr's
POSIX API Kconfig options and the naming for POSIX Options and
Option Groups.

The Kconfig option CONFIG_PTHREAD_IPC has been (ab)used for a
very long time for a variety of different purposes. However,
the standard Option / feature test macro for POSIX Threads is,
intuitively _POSIX_THREADS. There is a corresponding sysconf()
key named _SC_POSIX_THREADS.

Annoyingly, the POSIX Option Group that corresponds to the
Option is POSIX_THREADS_BASE, which is a minor inconsistency
in the standard.

The _POSIX_THREADS Option already includes mutexes, condition
variables, and thread-specific storage (keys). So with this
change, we also deprecate the redundant Kconfig variables that
do not have a corresponding match in the standard.

- CONFIG_PTHREAD_IPC
- CONFIG_PTHREAD
- CONFIG_PTHREAD_COND
- CONFIG_PTHREAD_MUTEX
- CONFIG_PTHREAD_KEY

Additionally, create Kconfig variables for those configurables
which we are lacking:

- CONFIG_POSIX_THREADS_EXT
- CONFIG_POSIX_THREAD_ATTR_STACKSIZE
- CONFIG_POSIX_THREAD_ATTR_STACKADDR
- CONFIG_POSIX_THREAD_PRIORITY_SCHEDULING
- CONFIG_POSIX_THREAD_PRIO_INHERIT
- CONFIG_POSIX_THREAD_PRIO_PROTECT
- CONFIG_POSIX_THREAD_SAFE_FUNCTIONS

Some Kconfig variables were renamed to more properly match the spec:

- CONFIG_MAX_PTHREAD_COUNT -> CONFIG_POSIX_THREAD_THREADS_MAX
- CONFIG_MAX_PTHREAD_KEY_COUNT -> CONFIG_POSIX_THREAD_KEYS_MAX

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

ycsin commented on lib/os/fdtable.c:
```diff
@@ -332,9 +329,8 @@ ssize_t write(int fd, const void *buf, size_t sz)
-FUNC_ALIAS(write, _write, ssize_t);
 
-int close(int fd)
+int zvfs_close(int fd)
```
are we going to group these `zvfs_*` functions into a file?

cfriedt replied:
That would be part of the ZVFS task - in this PR, I'm mainly concerned with correcting years of erroneous Kconfig options.

ycsin commented on lib/posix/options/Kconfig.barrier:
```diff
@@ -1,14 +1,27 @@
-type-function = pthread_barrier_wait
-rsource "Kconfig.template.pooled_ipc_type"
+menuconfig POSIX_BARRIERS
+    bool "POSIX barriers"
```
```suggestion
	bool "POSIX barriers"
```

ycsin commented on include/zephyr/posix/sys/utsname.h:
```diff
@@ -6,15 +6,23 @@
 struct utsname {
 	char sysname[sizeof("Zephyr")];
-	char nodename[CONFIG_POSIX_UNAME_NODENAME_LEN + 1];
+	char nodename[_UTSNAME_NODENAME_LENGTH + 1];
```
Probably can move the `+ 1` into the macro above as well?

cfriedt replied:
I'm kind of leaning toward making this PR as much of a straight text replacement as possible.

There are quite a few other minefields that need to be carefully traversed in terms of standard feature test macros, so I would rather not make drive-by changes if possible.

ycsin commented on lib/posix/options/Kconfig.deprecated:
```diff
@@ -56,6 +56,14 @@ config POSIX_MQUEUE
+	help
+	  This option is deprecated.
+
+	  Please use CONFIG_XSI_SYSTEM_LOGGING instead.
```
`XSI_SYSTEM_LOGGING` isnt defined in this commit

cfriedt replied:
I think I may have lost it in a rebase -- working on it..

cfriedt replied:
Fixed!

ycsin reviewed (commented):
I think this is a much needed clean up, thanks for this, some more comments.

Commit:
posix: relocate Kconfig.eventfd to Kconfig.compat

While event file descriptors (eventfd) are extremely useful, the
API has still not been formalized as part of the POSIX
specification.

Therefore, move this to a separate Kconfig menu for "compat"
(compatibility) features.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: deprecate POSIX_SIGNAL in favour of POSIX_SIGNALS

This change deprecates CONFIG_POSIX_SIGNAL in favour of
the plural CONFIG_POSIX_SIGNALS, which maps directly to the
name of the standard POSIX Option Group POSIX_SIGNALS.

Additionally, mark signals as experimental.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: reword aio and sched options. add experimental

The asynchronous I/O and (process) scheduling options needed
minor rewording.

Additionally, add the experimental flag to each of them since
they are not yet supported.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: move getpid() to Kconfig.procN and doc MULTI_PROCESS

Create a Kconfig option for POSIX_MULTI_PROCESS Option Group
and mark it as experimental.

Currently, Zephyr has two functions that belong to this
Option Group:
- sleep()
- getpid()

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
doc: posix: clean up some Kconfig options

Add links to POSIX-related Kconfig options that are related to
POSIX Options and Option Groups.

These include limits that might not have an associated standard
name and related Kconfig variables in other subsystems.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: net: update for POSIX_NETWORKING

Add a link to the Kconfig option CONFIG_POSIX_NETWORKING in
the documentation.

Additionally, add Kconfig options for other Network
dependencies in an effort to make POSIX Kconfig options
consistent with the specification.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
doc: posix: misc cleanup and add links to system interfaces

memlock, memlock range, shared memory objects, and synchronized
io were missing links.

Also, vdisable is '\0' (not 0).

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: create kconfig options for pse51, pse52, pse53

Create Kconfig "shortcuts" for PSE51, PSE52, and PSE53.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
doc: release: posix: deprecate non-normative POSIX Kconfig

Add release notes for the deprecation of a number of POSIX
Kconfig options. These have been deprecated so that we can
use more normative Kconfig variable identifiers, based on
the Options and Option Groups of IEEE 1003.1-2017.

To simplify migration, use

python $ZEPHYR_BASE/scripts/utils/migrate_posix_kconfigs.py \
    -r root_path

Additionally, document the removal of PTHREAD_BARRIER_DEFINE(),
EFD_IN_USE, EFD_FLAGS_SET, which were previously deprecated
>= 2 release cycles before.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

cfriedt commented on tests/posix/headers/src/unistd_h.c:
```diff
@@ -24,24 +24,23 @@ ZTEST(posix_headers, test_unistd_h)
-	/* zassert_not_equal(INT_MIN, _CS_POSIX_V7_THREADS_LDFLAGS); */ /* not implemented */
-	/* zassert_not_equal(INT_MIN, _CS_POSIX_V7_WIDTH_RESTRICTED_ENVS); */ /* not implemented */
-	/* zassert_not_equal(INT_MIN, _CS_V7_ENV); */ /* not implemented */
+	zassert_not_equal(INT_MIN, _CS_PATH);
```
It can be, but it's a separate commit, so not really a big deal.

cfriedt commented on lib/posix/options/sysconf.c:
```diff
@@ -4,263 +4,271 @@
+#include <zephyr/posix/sys/sysconf.h>
 #include <zephyr/posix/unistd.h>
 
+#ifdef CONFIG_POSIX_SYSCONF_IMPL_FULL
```
I think we just want to group this into `CONFIG_POSIX_SINGLE_PROCESS`, so having multiple conditions in CMake seems excessive.

cfriedt commented on lib/posix/options/sysconf.c:
```diff
@@ -4,263 +4,271 @@
 
+#ifdef CONFIG_POSIX_SYSCONF_IMPL_FULL
+
+#define z_sysconf(x) (long)CONCAT(__z_posix_sysconf, x)
```
`CONFIG_POSIX_SYSCONF` is actually not a standard feature test macro, so it needs to be removed eventually.

ycsin reviewed (commented):
🔥

Commit:
posix: eventfd: remove deprecated eventfd macros

Remove the EFD_IN_USE and EFD_FLAGS_SET macros as they were not
part of the public API.

Rename the internal versions, accordingly.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: eventfd: fix dependency cycle between net and posix

Until recently, the posix api was purely a consumer of the
network subsystem. However, a dependency cycle was added as
a stop-gap solution for challenges with the native platform.

Specifically,

1. eventfd symbols conflict with those of the host
2. eventfd was excluded from native libc builds via cmake

If any part of the posix were then to select the network
subsystem (which is a legitimate use case, given that networking
is a part of the posix api), we would get a build error due to
the Kconfig dependency cycle.

As usual, with dependency cycles, the cycle can be broken
via a third, mutual dependency.

What is the third mutual dependency? Naturally, it is ZVFS
which was planned some time ago. ZVFS will be where we
collect file-descriptor and FILE-pointer APIs so that we can
ensure consistency for Zephyr users.

This change deprecates EVENTFD_MAX in favour of
ZVFS_EVENTFD_MAX.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

jukkar commented on lib/posix/options/Kconfig.deprecated:
```diff
@@ -5,4 +5,13 @@
+
+config POSIX_LIMITS_RTSIG_MAX
+	int "_POSIX_RTSIG_MAX value in limits.h [DEPRECATED]"
+	default 0
```
Should we add `select DEPRECATED` here?

cfriedt replied:
Oddly, int options get a warning if they select deprecated, because they need to have a default value and are effectively always selected, AFAICT. I might be doing it wrong though. 

What I've done is add a build time `__WARN()` in select places if it is defined and non-zero.

ycsin replied:
I guess it isn't as straightforward to deprecate a Kconfig of type `int`:
https://github.com/zephyrproject-rtos/zephyr/blob/4b6d3f1e16ea8ccccbf85f1eaf7efd0caea11766/kernel/Kconfig.smp#L41-L46

cfriedt replied:
Agreed. It seems best practice would be to default the deprecated value to the non-deprecated value - those should be fixed now.

jukkar commented on doc/releases/release-notes-3.7.rst:
```diff
@@ -63,6 +63,10 @@ Deprecated in this release
+  * In efforts to use Kconfig options that map directly to the Options and Option Groups in
+    IEEE 1003.1-2017, the following Kconfig options have been deprecated (replaced by):
+
+    * :kconfig:option:`CONFIG_POSIX_LIMITS_RTSIG_MAX` (:kconfig:option:`CONFIG_POSIX_RTSIG_MAX`)
```
Suggestion to merge the changes in `doc: release: posix: deprecate CONFIG_POSIX_LIMITS_RTSIG_MAX` to the commit that does the actual deprecation.

cfriedt replied:
I wanted to keep release notes to separate commits.

What I could do is squash all of the changes to the release notes to 1 commit at the end.

cfriedt replied:
Fixed!

jukkar commented on doc/releases/release-notes-3.7.rst:
```diff
@@ -66,7 +66,13 @@ Deprecated in this release
   * In efforts to use Kconfig options that map directly to the Options and Option Groups in
     IEEE 1003.1-2017, the following Kconfig options have been deprecated (replaced by):
 
+    * :kconfig:option:`CONFIG_MAX_TIMER_COUNT` (:kconfig:option:`CONFIG_POSIX_TIMER_MAX`)
```
I suggest to merge all deprecation documentation commits to the actual commit that does the deprecation.

cfriedt replied:
I'm not crazy about having release notes commits mixed with code commits.

I could squash all of the release notes to 1 message though.

Does that work?

cfriedt replied:
Fixed!

jukkar commented on lib/posix/options/Kconfig.deprecated:
```diff
@@ -6,6 +6,27 @@
+config MAX_TIMER_COUNT
+	int "Maximum number of timer_t [DEPRECATED]"
+	default 0
+	help
```
Could you add `select DEPRECATED` to all deprecated kconfig options where it is missing.

jukkar commented on lib/posix/options/Kconfig.compat:
```diff
@@ -14,10 +17,16 @@ menuconfig EVENTFD
 	  be used as an event wait/notify mechanism together with POSIX calls
 	  like read, write and poll.
 
+if EVENTFD
```
Does not harm and can be certainly left here, but as the `EVENTFD_MAX` has `depends on EVENTFD` this `if EVENTFD` is not needed.

cfriedt replied:
Fixed!

jukkar commented on lib/posix/options/Kconfig.net:
```diff
@@ -2,10 +2,48 @@
+	  https://pubs.opengroup.org/onlinepubs/9699919799/xrat/V4_port.html#tag_24_03_04
+
+config POSIX_RAW_SOCKETS
+	bool "POSIX IPv6 support"
```
Typo, should probably be "POSIX RAW socket support" or similar

cfriedt replied:
Fixed!

jukkar commented on lib/os/zvfs/Kconfig:
```diff
@@ -0,0 +1,35 @@
+
+menuconfig ZVFS
+	bool "Zephyr virtual filesystem (ZVFS) support [EXPERIMENTAL]"
+	select FDTABLE
```
Probably need to `select EXPERIMENTAL` here

cfriedt replied:
Fixed!

jukkar reviewed (commented):
Lot of commits and I probably missed lot of stuff so I will return to this in future versions.

cfriedt commented on tests/posix/headers/prj.conf:
```diff
@@ -9,6 +9,7 @@ CONFIG_NET_TEST=y
 CONFIG_TEST_RANDOM_GENERATOR=y
 CONFIG_NET_SOCKETS=y
 CONFIG_NETWORKING=y
+CONFIG_NET_SOCKETS_PACKET=y
```
Some Kconfig options were added near the latest commit (update to POSIX_NETWORKING or pse51, pse52, ...) because those had minor modifications on the behaviour of POSIX_API.

Previously, we used POSIX_API as a kind of "enable all". The problem was, the feature set kept growing. Not only did binaries get bigger, but some tests would fail to fit into rom / ram. Then I tried to make POSIX_API equal to BASE but too many things broke. Eventually, I determined that generally, 99% of tests and samples would build with the current interpretation, which is actually a fixed set of features.

cfriedt replied:
Fixed

cfriedt commented on lib/posix/options/rwlock.c:
```diff
@@ -34,16 +34,16 @@ LOG_MODULE_REGISTER(pthread_rwlock, CONFIG_PTHREAD_RWLOCK_LOG_LEVEL);
  * perspective of the application). With a linear space, this means that
  * the theoretical pthread_rwlock_t range is [0,2147483647].
  */
-BUILD_ASSERT(CONFIG_MAX_PTHREAD_RWLOCK_COUNT < PTHREAD_OBJ_MASK_INIT,
```
I don't believe this one does need to be deprecated. There is no better alternative in the spec for [unistd.h](https://pubs.opengroup.org/onlinepubs/7908799/xsh/unistd.h.html) or [limits.h](https://pubs.opengroup.org/onlinepubs/009695399/basedefs/limits.h.html) fwict.

dleach02 commented on doc/releases/release-notes-3.7.rst:
```diff
@@ -60,6 +60,50 @@ Deprecated in this release
 
+ * POSIX API
+
+  * Deprecated :c:macro:`PTHREAD_BARRIER_DEFINE` has been removed.
```
isn't this stuff that should go into the migration guide?

cfriedt replied:
I could technically add a script that would adjust .conf files automatically.

That would be a good thing for the migrations n guide

cfriedt replied:
Fixed!

cfriedt commented:
Seems like a lot of broken unrelated code in mbedtls. I rebased.

aescolar commented:
Seems this PR hit the limit of reviewers and failed to notify a lot of affected areas maintainers/collaborators. Notifying them here:
@57300
@aescolar
@andyross
@avisconti
@BeckmaR
@dcpleung
@erwango
@jakub-uC
@jeremybettis
@jfischer-no
@jukkar
@kartben
@keith-packard
@kludentwo
@krish2718
@MaureenHelm
@nashif
@nordicjm
@nordic-krch
@pdgendt
@peter-mitsis
@rerickson1
@rlubos
@SeppoTakalo
@ssharks
@stephanosio
@tbursztyka
@teburd
@tejlmand
@tristan-google
@ubieda
@vaishnavachath
@vanti
@ycsin
@yperess

aescolar commented on samples/subsys/shell/shell_module/prj.conf:
```diff
@@ -9,7 +9,8 @@ CONFIG_THREAD_MONITOR=y
 CONFIG_THREAD_NAME=y
 CONFIG_DEVICE_SHELL=y
-CONFIG_POSIX_CLOCK=y
+CONFIG_POSIX_API=y
```
With these changes, these tests fail to build for native_posix[//64], so they will fail in main's CI.
Please either ensure they keep working, or filter these platforms out.

cfriedt replied:
Certainly!

I'll wait to push changes until later, just to ensure that other change requests have been captured as well.

cfriedt replied:
Fixed!

aescolar commented on scripts/utils/migrate_posix_kconfigs.py:
```diff
@@ -0,0 +1,134 @@
+"""
+Utility script to migrate Zephyr-based projects to new POSIX Kconfig options
```
Minor: It would be helpful to indicate here in which Zephyr version the rename happened. Both to guide users and so that later is clearer when this transitional script could be removed.

cfriedt replied:
Fixed!

jukkar reviewed (dismissed):
I have to say the amount of changes is overwhelming and difficult to review and I probably missed something. Anyway, looks reasonable to me.

andyross reviewed (dismissed):
No arguments with the change, can't see anything I normally care about that would break.   Absolutely don't claim to have reviewed every line, but feel free to assume further approvals from me as this evolves.

ycsin reviewed (dismissed):
💯

aescolar commented on scripts/utils/migrate_posix_kconfigs.py:
```diff
@@ -0,0 +1,137 @@
+"""
+Utility script to migrate Zephyr-based projects to normative POSIX Kconfig options.
+
+This script should be used for migrating from verions of Zephyr newer than v3.7.0 to Zephyr
+verions v3.7.0 or later.
```
```suggestion
This script should be used for migrating from versions of Zephyr older than v3.7.0 to Zephyr
version v3.7.0 or later.
```

cfriedt replied:
Fixed.

cfriedt commented:
I'm going to go out on a limb and say that unrelated error is unrelated.

https://github.com/zephyrproject-rtos/zephyr/actions/runs/9329767331/job/25682654042?pr=73047

It fails on main too and is bisectable.

Bad commit is cf4495b196051d7cf634421eb79c26210162f4be . Hotfix is in https://github.com/zephyrproject-rtos/zephyr/pull/73604
cc @dleach02

cfriedt commented:
rebased after https://github.com/zephyrproject-rtos/zephyr/pull/73604 was merged

cfriedt commented:
@aescolar - your change requests have been made. Is there anything else?

ycsin reviewed (approved):
Refresh +1

cfriedt commented:
Friendly ping to @peter-mitsis @andyross @jukkar . Can I solicit more refresh +1s?

andyross reviewed (approved):
Love the standard names.  Not a big fan of renaming churn in general (makes bisection across long history a high impedance process because you need to remember what the conventions were at each point), but if we're going to do it "all in a single PR" is for sure the best way.

SeppoTakalo commented:
The Zephyr 3.7 migration guide still has this:
> * The DNS resolver and mDNS/LLMNR responders are converted to use socket service API.
  This means that the number of pollable sockets in the system might need to be increased.
  Please check that the values of :kconfig:option:`CONFIG_NET_SOCKETS_POLL_MAX` and
  :kconfig:option:`CONFIG_POSIX_MAX_FDS` are high enough. Unfortunately no exact values
  for these can be given as it depends on application needs and usage. (:github:`72834`)

should it be refactored so you don't need to recursively read the migration guide?

fabiobaltieri commented on include/zephyr/posix/pthread.h:
```diff
@@ -153,7 +153,7 @@ int pthread_condattr_setclock(pthread_condattr_t *att, clockid_t clock_id);
  * Initialize a rwlock with the default rwlock attributes.
  */
-#define PTHREAD_RWLOCK_INITIALIZER (-1)
+#define POSIX_RWLOCK_INITIALIZER (-1)
```
hey @cfriedt is this correct? I bumped into a library that was using `PTHREAD_RWLOCK_INITIALIZER` (boringssl) and now does not build anymore, I see that `/usr/include/pthread.h` on my system defines `PTHREAD_RWLOCK_INITIALIZER` as well

cfriedt replied:
@fabiobaltieri - good find. `PTHREAD_RWLOCK_INITIALIZER` is the standard macro name. Are you good to submit a fix?

fabiobaltieri replied:
yeah I can do that

fabiobaltieri replied:
there you go https://github.com/zephyrproject-rtos/zephyr/pull/73808

fabiobaltieri commented:
Another issue I bumped into this: the sysconfig change somehow collides with a macro in the stm32 hals, try:

`west build -p -b nucleo_f410rb samples/posix/uname/ -DCONFIG_POSIX_SYSCONF_IMPL_FULL=y`

```
/home/fabiobaltieri/zephyrproject/modules/hal/stm32/stm32cube/stm32f4xx/drivers/include/Legacy/stm32_hal_legacy.h:475:39: error: 'FLASH_PAGE_SIZE' undeclared (first use in this function); did you mean 'FLASH_CR_PSIZE'?
  475 | #define PAGESIZE                      FLASH_PAGE_SIZE
```

I think there's a macro collision for `PAGESIZE`, not sure what's the proper fix though

@cfriedt @erwango

cfriedt commented:
> Another issue I bumped into this: the sysconfig change somehow collides with a macro in the stm32 hals, try:

@fabiobaltieri - I believe this was fixed a while ago by Erwan. Did you `west update`?

cfriedt commented:
> The Zephyr 3.7 migration guide still has this:
> 
> > * The DNS resolver and mDNS/LLMNR responders are converted to use socket service API.
> >   This means that the number of pollable sockets in the system might need to be increased.
> >   Please check that the values of :kconfig:option:`CONFIG_NET_SOCKETS_POLL_MAX` and
> >   :kconfig:option:`CONFIG_POSIX_MAX_FDS` are high enough. Unfortunately no exact values
> >   for these can be given as it depends on application needs and usage. (:github:`72834`)
> 
> should it be refactored so you don't need to recursively read the migration guide?

Probably, yea. @SeppoTakalo are you good to submit a fix?

fabiobaltieri commented:
> @fabiobaltieri - I believe this was fixed a while ago by Erwan. Did you `west update`?

```shell
~/zephyrproject/zephyr$ git fetch
~/zephyrproject/zephyr$ git checkout --detach main
HEAD is now at d3081e2f30f net: lwm2m: On write, use server selected block size
~/zephyrproject/zephyr$ west update
~/zephyrproject/zephyr$ west build -p -b nucleo_f410rb samples/posix/uname/ -DCONFIG_POSIX_SYSCONF_IMPL_FULL=y
<broken>
```

Looks like it's still a problem, maybe manifest is out of date?

cfriedt commented:
> Looks like it's still a problem, maybe manifest is out of date?

Need to update the manifest with https://github.com/zephyrproject-rtos/hal_stm32/pull/209

fabiobaltieri commented:
> Need to update the manifest with [zephyrproject-rtos/hal_stm32#209](https://github.com/zephyrproject-rtos/hal_stm32/pull/209)

Ah, I do have that commit, but that just removes the path from the include path list, the header is still being included:

```
In file included from /home/fabiobaltieri/zephyrproject/modules/hal/stm32/stm32cube/stm32f4xx/drivers/include/stm32f4xx_hal_def.h:30,
                 from /home/fabiobaltieri/zephyrproject/modules/hal/stm32/stm32cube/stm32f4xx/drivers/include/stm32f4xx_hal_rcc.h:27,
                 from /home/fabiobaltieri/zephyrproject/modules/hal/stm32/stm32cube/stm32f4xx/drivers/include/stm32f4xx_hal_conf.h:280,
                 from /home/fabiobaltieri/zephyrproject/modules/hal/stm32/stm32cube/stm32f4xx/drivers/include/stm32f4xx_hal.h:29,
                 from /home/fabiobaltieri/zephyrproject/modules/hal/stm32/stm32cube/stm32f4xx/soc/stm32f4xx.h:287,
                 from /home/fabiobaltieri/zephyrproject/zephyr/soc/st/stm32/stm32f4x/./soc.h:23,
                 from /home/fabiobaltieri/zephyrproject/zephyr/modules/cmsis/./cmsis_core_m.h:24,
                 from /home/fabiobaltieri/zephyrproject/zephyr/modules/cmsis/./cmsis_core.h:10,
                 from /home/fabiobaltieri/zephyrproject/zephyr/include/zephyr/arch/arm/asm_inline_gcc.h:24,
                 from /home/fabiobaltieri/zephyrproject/zephyr/include/zephyr/arch/arm/asm_inline.h:18,
                 from /home/fabiobaltieri/zephyrproject/zephyr/include/zephyr/arch/arm/arch.h:33,
                 from /home/fabiobaltieri/zephyrproject/zephyr/include/zephyr/arch/cpu.h:19,
                 from /home/fabiobaltieri/zephyrproject/zephyr/include/zephyr/kernel_includes.h:36,
                 from /home/fabiobaltieri/zephyrproject/zephyr/include/zephyr/kernel.h:17,
                 from /home/fabiobaltieri/zephyrproject/zephyr/arch/arm/core/offsets/offsets_aarch32.c:28,
                 from /home/fabiobaltieri/zephyrproject/zephyr/arch/arm/core/offsets/offsets.c:9:
```

@erwango could you look into it?
